PCI-DSS centric benchmarks for RHEL6 and 7

mpreisler opened the pull request:
Hi,
just opening a discussion after I have progressed with the PCI-DSS centric Benchmark feature. I have code that can build PCI-DSS Benchmarks centered around PCI-DSS requirements instead of our own hierarchy. This is all generated from references we already have in the benchmarks. In essence this transformation reverses the mapping from rules->pci-dss requirements to pci-dss requirements->rules. The situation where one rule references multiple pci-dss requirements is handled by duplicating the rule with unique ID.

This is how the PCI-DSS benchmark looks in SCAP Workbench while tailoring:
![image](https://cloud.githubusercontent.com/assets/753153/15157969/ddec8226-16bb-11e6-9812-c020b0b83f7b.png)

This is how the HTML guide looks:
![image](https://cloud.githubusercontent.com/assets/753153/15158024/10dac4fe-16bc-11e6-9831-3f2f5c5f74b7.png)

I will paste how the HTML report looks like tomorrow.

Q&A:
- could we add all this into the ssg-rhel7-ds.xml file and not have a separate sds? I would love to do this but the SSG tooling and other OpenSCAP tools don't fully support this.
- could we add more description text from the PCI-DSS PDF? Probably but it requires clever code parsing the PDF itself. What I have done is fairly simplistic. And only adds the first line.
- the transformations are licensed under LGPL. Does this affect the SSG policies? No, it's just the transformation that's LGPL, not the policies.

Open questions:
- should I filter out all profiles except pcidss when building the pcidss benchmark? This would save time at build, less guides would need to be generated.
- how do we scale this in the future? If we want something similar for another profile we can't keep adding more datastreams.

mpreisler commented:
[retest]

iankko commented:
@mpreisler 

Hi Martin, first thank you for this effort!

> Q&A:
> could we add all this into the ssg-rhel7-ds.xml file and not have a separate sds?
I would love to do this but the SSG tooling and other OpenSCAP tools don't fully support this.

Can you elaborate which SSG tooling part doesn't support this? The issues I expect with having two RHEL-7 benchmarks being:
* it will break the UX experience (users are now accustomed to find all available profiles in one datastream file),
* it will break the current tooling experience (for the tools using SSG for their functionality - Oscap Anaconda Addon, Satellite 5/6, ManageIQ in the future). Having two separate datastream files we would need to notify these users / products they need to modify their codebases to start counting with the fact the former ```ssg-rhel7-ds.xml``` is now split into two files.

>    could we add more description text from the PCI-DSS PDF? Probably but it requires
clever code parsing the PDF itself. What I have done is fairly simplistic. And only adds the first line.

I would vote for adding whole prose from the PCI DSS v3.1 standard PDF. And use just OVAL checks and remediations from the SSG content. This will IMHO allow the auditors to approach the experience they have while scanning PCI DSS with different tools than OpenSCAP.

>    the transformations are licensed under LGPL. Does this affect the SSG policies?
No, it's just the transformation that's LGPL, not the policies.

We could change the SSG license from Public Domain to Public Domain + LGPL (keep PD for the content itself, and add LGPG for the transformation itself). This is commonly used scenario across different packages / software products used in Linux distributions AFAICT.

> Open questions:
>    should I filter out all profiles except pcidss when building the pcidss benchmark?
This would save time at build, less guides would need to be generated.

I would say yes. On the other hand if we are gonna to go the way of having two separate benchmarks, we could probably keep the current ```ssg-rhel7-ds.xml``` as it is now (IOW containing all profiles), with explanation that this PCI DSS benchmark is just for users trying to prescan their infrastructure for possible PCI DSS compliance (mainly to find out what needs to be corrected, when PCI DSS compliance is attempted). The second ```ssg-rhel7-pcidss-auditor-ds.xml``` (current ```ssg-rhel7-pcidss-ds.xml``` in your code) benchmark would contain just PCI DSS profile, and would have the layout as requested by PCI DSS v3.1 standard PDF file.

> how do we scale this in the future? If we want something similar for another profile
we can't keep adding more datastreams.

That's correct. Not to speak to abstract, there actually exists such a request already:
[*]  https://github.com/OpenSCAP/scap-security-guide/issues/805

I am afraid approving the above proposed two datastreams approach might induce that unpleasant precedent when people would use the PCI DSS approach as an argument why their desired profile should be also put into separate datastream (speaking about [*] it would be the case for the USGCB profile).

We can't keep adding more datastreams because the other tooling located on the top of SSG (OAA, Satellite, ManageIQ) and also can't keep adding them, because the convenient method to scan the system against different profiles would disappear in the moment, we split profiles per different datastreams.

Which returns us to the first point again - can you elaborate which specific tooling (SSG, oscap) issues are preventing to keep having one ```ssg-rhel7-ds.xml``` datastream file, but having multiple XCCDF components (first formatted in "traditional SSG" way, while the second formatted in more "PCI DSS" like way)?

mpreisler commented:
> I would vote for adding whole prose from the PCI DSS v3.1 standard PDF. And use just OVAL checks and remediations from the SSG content. This will IMHO allow the auditors to approach the experience they have while scanning PCI DSS with different tools than OpenSCAP.

I certainly vote for that as well but the code isn't trivial. The reason why we don't do it now is not because I decided not to do it, it's because it's really hard to implement it.

> Which returns us to the first point again - can you elaborate which specific tooling (SSG, oscap) issues are preventing to keep having one ssg-rhel7-ds.xml datastream file, but having multiple XCCDF components (first formatted in "traditional SSG" way, while the second formatted in more "PCI DSS" like way)?

- `oscap xccdf generate guide` in OpenSCAP
- `build-all-html-guides.py` in SSG
- `enable derivatives script` in SSG
- SCAP Workbench has a the assumption that there is just one benchmark in the datastream for usability purposes, this can be changed easily but right now it's there
- I'd have to look into Foreman SCAP integration again to be sure but last I checked you can't choose a component ID there conveniently
- Satellite 5, outside of hacking the command line arguments you can't choose a component ID

We can put the feature on hold here, fix all this and then go with the one datastream approach. I wanted to discuss this here.

redhatrises commented:
>> I would vote for adding whole prose from the PCI DSS v3.1 standard PDF. And use just OVAL checks and remediations from the SSG content. This will IMHO allow the auditors to approach the experience they have while scanning PCI DSS with different tools than OpenSCAP.

> I certainly vote for that as well but the code isn't trivial. The reason why we don't do it now is not because I decided not to do it, it's because it's really hard to implement it.

@mpreisler I agree with @iankko. We definitely shouldn't be using more than one datastream file as we would be opening up a not only a potential maintenance nightmare but a "why can't product x have its own datastream" situation as well.

>> Which returns us to the first point again - can you elaborate which specific tooling (SSG, oscap) issues are preventing to keep having one ssg-rhel7-ds.xml datastream file, but having multiple XCCDF components (first formatted in "traditional SSG" way, while the second formatted in more "PCI DSS" like way)?

>- `oscap xccdf generate guide` in OpenSCAP
- `build-all-html-guides.py` in SSG
- `enable derivatives script` in SSG
- SCAP Workbench has a the assumption that there is just one benchmark in the datastream for usability purposes, this can be changed easily but right now it's there
- I'd have to look into Foreman SCAP integration again to be sure but last I checked you can't choose a component ID there conveniently
- Satellite 5, outside of hacking the command line arguments you can't choose a component ID

> We can put the feature on hold here, fix all this and then go with the one datastream approach. I wanted to discuss this here.

While it may be harder, let's open up tickets for tracking and start working through the issues.

mpreisler commented:
Alright, we seem to have a consensus here: I will change the PR to build only one SDS with 2 benchmarks in it. I will file RFEs to everything that has problems with the new datastreams.

mpreisler commented:
Changed the build. Now it builds one datastream for each product with 2 checklists inside. There are a couple of issues that need to be resolved.

- `build-all-guides.py` needs to be updated to handle SDS with multiple datastreams.
- oscap needs to be patched, right now it adds everything twice, the OVAL files, OCIL files, etc... This is not a big issue right now but we need to fix it

I am having problems getting SCAP Workbench to work with this datastream, I need to look into this. Will update this ticket.

mpreisler commented:
This is the datastream it currently outputs for RHEL7.

[ssg-rhel7-ds.xml.zip](https://github.com/OpenSCAP/scap-security-guide/files/259711/ssg-rhel7-ds.xml.zip)

iankko commented:
@mpreisler 
Thank you for the preview datastream. Ad
> I am having problems getting SCAP Workbench to work with this datastream, I need to look into this. Will update this ticket.

Is the issue you mean above when changing DS component, list of profiles isn't refreshed (even when selecting the second component, there are still all profiles listed). Or something else?

mpreisler commented:
> Is the issue you mean above when changing DS component, list of profiles isn't refreshed (even when selecting the second component, there are still all profiles listed). Or something else?

They are refreshed but the original profiles show up instead of the new ones. I need to debug this. I hope it's an issue in SCAP Workbench and not OpenSCAP itself.

mpreisler commented:
I have fixed the invalid selectors issue.

Issues to solve here:
- `oscap ds sds-add` adds the RHEL7 OVAL 3 times for some reason. It should deduplicate. At least it should avoid using the same ID twice
- scap-workbench doesn't use the hierarchy of the right benchmark
- `oscap xccdf generate guide --xccdf-id scap_org.open-scap_cref_output--ssg-rhel7-pcidss-xccdf-1.2.xml --profile xccdf_org.ssgproject.content_profile_pci-dss ssg-rhel7-ds.xml > o.html` doesn't use the right hierarchy
- `build-all-guides.py` refuses to continue when it finds 2 XCCDF Benchmarks in one file

Updated DS: [ssg-rhel7-ds.xml.zip](https://github.com/OpenSCAP/scap-security-guide/files/262229/ssg-rhel7-ds.xml.zip)

mpreisler commented:
OpenSCAP can now generate guide of datastreams with multiple Benchmarks, see https://github.com/OpenSCAP/openscap/commit/a5bda101e07f9383e14109270df4748fb12ea4e5

mpreisler commented:
[test this]

redhatrises commented:
[test this]

mpreisler commented:
Fixed the issue with SCAP Workbench in https://github.com/OpenSCAP/scap-workbench/commit/b9bc1f7259ca971c3402b3470318037ff6f39adc

The issue fortunately wasn't in OpenSCAP itself. SCAP Workbench can now select and tailor the right benchmark. I haven't found other issues.

mpreisler commented:
> Moving to 0.1.31 (couple of things to be resolved in this one + bigger change better to stay couple of days in repo prior making new release).

Agreed.

I have just rebased this to prevent too many conflicts piling up.

mpreisler commented:
Managed to get a little bit more work done on this. Now we can generate guides even from SDS with multiple benchmarks in them.

mpreisler commented:
[test this please]

mpreisler commented:
I have sneaked 2 small improvements in build-all-guides.py in this PR, it was an accident but I think I can leave them here so they get merged eventually.

mpreisler commented:
I think I have fixed the validation issues but we'll see what Jenkins says :-)

mpreisler commented:
This is how the oscap info output looks like with this PR. There are 2 duplications yet to be solved:
- `scap_org.open-scap_cref_ssg-rhel7-oval.xml` and `scap_org.open-scap_cref_output--ssg-rhel7-oval.xml`
- `scap_org.open-scap_cref_ssg-rhel7-ocil.xml` and `scap_org.open-scap_cref_output--ssg-rhel7-ocil.xml`

These are not affecting use cases, they just bloat the datastream. I propose to get rid of them using something like `oscap ds deduplicate ssg-rhel7-ds.xml` in the future.

```
$ oscap info ssg-rhel7-ds.xml 
Document type: Source Data Stream
Imported: 2016-07-21T15:33:34

Stream: scap_org.open-scap_datastream_from_xccdf_ssg-rhel7-xccdf-1.2.xml
Generated: (null)
Version: 1.2
Checklists:
	Ref-Id: scap_org.open-scap_cref_ssg-rhel7-xccdf-1.2.xml
		Status: draft
		Generated: 2016-07-21
		Resolved: true
		Profiles:
			xccdf_org.ssgproject.content_profile_standard
			xccdf_org.ssgproject.content_profile_pci-dss
			xccdf_org.ssgproject.content_profile_C2S
			xccdf_org.ssgproject.content_profile_rht-ccp
			xccdf_org.ssgproject.content_profile_common
			xccdf_org.ssgproject.content_profile_stig-rhel7-workstation-upstream
			xccdf_org.ssgproject.content_profile_stig-rhel7-server-gui-upstream
			xccdf_org.ssgproject.content_profile_stig-rhel7-server-upstream
			xccdf_org.ssgproject.content_profile_ospp-rhel7-server
			xccdf_org.ssgproject.content_profile_nist-cl-il-al
			xccdf_org.ssgproject.content_profile_cjis-rhel7-server
		Referenced check files:
			ssg-rhel7-oval.xml
				system: http://oval.mitre.org/XMLSchema/oval-definitions-5
			ssg-rhel7-ocil.xml
				system: http://scap.nist.gov/schema/ocil/2
			http://www.redhat.com/security/data/oval/com.redhat.rhsa-RHEL7.xml.bz2
				system: http://oval.mitre.org/XMLSchema/oval-definitions-5
	Ref-Id: scap_org.open-scap_cref_output--ssg-rhel7-pcidss-xccdf-1.2.xml
		Status: draft
		Generated: 2016-07-21
		Resolved: true
		Profiles:
			xccdf_org.ssgproject.content_profile_pci-dss
		Referenced check files:
			ssg-rhel7-oval.xml
				system: http://oval.mitre.org/XMLSchema/oval-definitions-5
			ssg-rhel7-ocil.xml
				system: http://scap.nist.gov/schema/ocil/2
			http://www.redhat.com/security/data/oval/com.redhat.rhsa-RHEL7.xml.bz2
				system: http://oval.mitre.org/XMLSchema/oval-definitions-5
Checks:
	Ref-Id: scap_org.open-scap_cref_ssg-rhel7-oval.xml
	Ref-Id: scap_org.open-scap_cref_ssg-rhel7-ocil.xml
	Ref-Id: scap_org.open-scap_cref_output--ssg-rhel7-oval.xml
	Ref-Id: scap_org.open-scap_cref_output--ssg-rhel7-ocil.xml
	Ref-Id: scap_org.open-scap_cref_output--ssg-rhel7-cpe-oval.xml
Dictionaries:
	Ref-Id: scap_org.open-scap_cref_output--ssg-rhel7-cpe-dictionary.xml
```

mpreisler commented:
Needed one more fix to build on RHEL6, now it should pass... fingers crossed!

mpreisler commented:
Argh, more issues on RHEL6, hopefully fixed now.

mpreisler commented:
@iankko I may need your help here with the move OCIL script. I don't understand some parts of it. For example there is a condition that sets some variables that are then used out of that scope. Could be that it's a bug, could be that I am misunderstanding something really badly.

iankko commented:
@mpreisler 

Hi Martin,

  thanks for checking. 

> @iankko I may need your help here with the move OCIL script. I don't understand some parts of it. For example there is a condition that sets some variables that are then used out of that scope. Could be that it's a bug, could be that I am misunderstanding something really badly.
> EDIT: reference https://github.com/mpreisler/scap-security-guide/blob/b6eec94537b77d4cc6aa1bf8268d21ddf42d6c7f/shared/transforms/datastream_move_ocil_to_ds_checks.py#L107

The check for the count of OCIL components in the DS was added just like sanity check (to ensure there's just one OCIL component in the DS). But since your change is allowing / generating multiple XCCDFs in one DataStream, it's pretty likely that check will fail (IOW there will be more OCILs and that check will fail).

But looking at the current failure:
  https://jenkins.open-scap.org/job/scap-security-guide-pull-requests/453/

seems the issue is missing 'timestamp' attribute for some of the OCIL components (the default 'None' value is used, since it failed to obtain the actual timestamp attribute from the component). The point here was to preserve the existing timestamp from the component (rather to generate a new one), and re-use that timestamp while moving the component.

I would look (via some verbose prints) which components it's actually operating on, and why there isn't a timestamp for it.

For the count of OCIL components in DS check, that one can be probably removed (since AFAIK the SSG build system creates one OCIL per XCCDF via XSLT transformations). So once more XCCDFs are present in DS, there will be more OCILs too (and that check will complain).

mpreisler commented:
It seems I have finally fixed the move OCIL script... fingers crossed for the 4th time..

mpreisler commented:
This is now ready for review & merge. @iankko @redhatrises @jan-cerny @ybznek

redhatrises commented:
@mpreisler 

- Remind me why `shared/transforms/pcidss/PCI_DSS_v3-1.pdf` instead of `shared/references/PCI_DSS_v3-1.pdf`?
- Also why are we putting transforms under a specific content directory i.e. `shared/transforms/pcidss` instead of putting all transforms under `shared/transforms`? It seems like this could create extra work in the future for any new content. Although, I guess you could argue it is cleaner.

mpreisler commented:
> Remind me why shared/transforms/pcidss/PCI_DSS_v3-1.pdf instead of shared/references/PCI_DSS_v3-1.pdf?
> Also why are we putting transforms under a specific content directory i.e. shared/transforms/pcidss instead of putting all transforms under shared/transforms? It seems like this could create extra work in the future for any new or existing content. Although, I guess you could argue it is cleaner.

I had it in `shared/transforms` first but since multiple files are involved and multiple steps I have put it in a special directory. The PCI-DSS PDF being put there is related to that.

The workflow is that you download fresh PCI-DSS PDF, run a script that mines ID from it and that script generates a json file with the PCI-DSS hierarchy. This json file is committed and later used to generate the PCI-DSS benchmark.

redhatrises commented:
@mpreisler is a separate `pcidss` directory needed under `shared/transforms`?

mpreisler commented:
> @mpreisler is a separate pcidss directory needed under shared/transforms?

It's not needed but it's preferable to avoid polluting shared/transforms with PDF and json files. I can move it but I think its' better if it's in a separate directory.

ybznek commented on shared/transforms/pcidss/generate_pcidss_json.py:
Great idea :+1:

ybznek commented on .gitignore:
```diff
@@ -89,3 +89,6 @@ shared/oval/*.ini
 shared/output/*.xml
+
+# Ignore PCI_DSS_v3-1.pdf
+shared/transforms/pcidss/PCI_DSS_v3-1.pdf
```
Imho `shared/transforms/pcidss/*.pdf` is more useful, but it is not very important

mpreisler commented:
Fixed the HTML guide index issues, thanks @ybznek for reporting.

ybznek commented:
@mpreisler 

My another/not very important hints:
* have you noticed that PCI_DSS_v3-2 is available? If we want to use it, we should deal with it it another PR.
* table of content is confusing, because number of requirements don't match numbers of "chapters" 

I think you have done great piece of work! after your *ack* I would merge it.

iankko commented:
Looks good enough for me for the initial version. ACK (to be enhanced in future PRs / versions).

iankko commented:
> Looks good enough for me for the initial version. ACK (to be enhanced in future PRs / versions).

To follow up on this - when the code is in, more people can look, test, report and fix the issues (than Martin to need to fix all of them).
